README
ResourcePool - A connection caching and pooling framework

Copyright (c) 2001-2009 by Markus Winand <mws@fatalmind.com>

See COPYRIGHT section in the ResourcePool(3pm) manpage.

Recent Changes
--------------
  *  ResourcePool did use a stringified object as key in a hash 
     reference. This has been replaced by using Tie::RefHash. 

Installation
------------
    ResourcePool is a pure perl implemented module and should 
    therefore run everywhere where perl is already installed.

    The well known procedure is:
        perl Makefile.PL
        make
        make test
        make install 

Examples
--------
    Are included in the manpages which come with this distribution.
    The ResourcePool::BigPicture(3pm) manpage is the right point 
    to start to get the overview, after that feel free to visit
    the ResourcePool(3pm) and ResourcePool::LoadBalancer(3pm)
    manpages.

Resource bindings
-----------------
    To do something usefull with ResourcePool you will need 
    resource bindings. There are some bindings available for
    download from CPAN. Those bindings live in the
    ResourcePool::Resource namespace, for example the Net::LDAP 
    binding has the name ResourcePool::Resource::Net::LDAP, if you 
    can not find the required bindings, please consult the 
    ResourcePool::ExtensionGuide(3pm) manpage for details how to 
    create one.

    Available bindings on CPAN (as of 2009-11-25):
       Net::LDAP
       DBI
       Alzabo
       SOAP::Lite

    The Net::LDAP and DBI bindings where included in the ResourcePool
    core package in all 0.99xx versions of ResourcePool. Since
    release 1.0000 they are no longer included in this package
    and have to be installed seperately.

How to contribute?
------------------
    Submit patches for bugs you discovered (or report the at least), 
    upload your own resource bindings to CPAN. 

Authors contact
---------------
       e-mail: mws@fatalmind.com
     homepage: http://www.fatalmind.com/
    whishlist: http://www.amazon.de/exec/obidos/wishlist/1JW6LC92SAK62/
               http://amazon.com/gp/registry/3BXRJP9W3WMDX


Changes
Changelog for ResourcePool

V1.0105 - 2009-11-25

  *   Made the timeout in the tests less restritive to avoid false 
      failure in case the test is performad on a loaded machine (Bug #29200).

V1.0104 - 2003-05-07

  *  ResourcePool did use a stringified object as key in a hash 
     reference. This has been replaced by using Tie::RefHash. 

V1.0103 - 2003-05-07

  *  Applied a patch from Chad Redman which is basically the same as 
     the patch from Eric Parusel which was partitially applied to 
     1.0102. This time I applied the patch to the correct CVS branch 
     *argl*. 

  *  Minor corrections and improvements in documentation. 

V1.0102 - 2003-03-27

  *  Two new LoadBalancer policies have been introduced: FailBack is 
     a alias for FallBack, and FailOver which acts like FailBack but 
     does stay on the fall back node. 

  *  Applied patch from Eric Parusel to fix a problem when debuging 
     LoadBalancer. 

V1.0101 - 2003-03-14

  *  An obsolete dependency for the Storable module has been removed. 

V1.0100 - 2003-02-27

  *  This release introduces a new way to use ResourcePool and 
     LoadBalancer. It implements the Command design pattern as 
     described in the GOF Design Patterns book. This release is a 
     preview, the features are not yet fully implemented and might 
     therefore not sufficient for you needs right now. 

V1.0000 - 2003-01-02

  *  From this release on, the ResourcePool package will not contain 
     resource bindings anymore. The previously included bindings for 
     DBI and Net::LDAP are available in the ResourcePool-Resource-DBI 
     and ResourcePool-Resource-Net-LDAP packages on CPAN. 

  *  Applied patch from James G Smith which introduced the is_created 
     method to ResourcePool::Singleton. This does not change the 
     behavior of any ResourcePool module. 

V0.9910 - 2002-11-25

  *  The runtime overhead introduced by ResourcePool and LoadBalancer 
     has been reduced. Depending on you configuration this release 
     has up to 35% less overhead. 

  *  Fixed a bug in LoadBalancer's documentation. The Weight 
     parameter works only for the LeastUsage policy. 

  *  A general code cleanup has been done to gain the performance 
     benefits. 

  *  Added tests for LoadBalancer. 

V0.9909 - 2002-10-07

  *  A refactoring of the Singleton behavior of ResourcePool::Factory 
     makes it possible to implement nicer factories 

  *  Adopted ResourcePool::Factory::Net::LDAP to have a more native 
     interface. 

  *  Fixed links within the POD documentation (was broken for some 
     pod2html translators) 

  *  Added documentation for Net::LDAP bindings which were lost in 
     the previous release. 

V0.9908 - 2002-09-30

  *  Added documentation: ResourcePool::BigPicture, ResourcePool::UML 
     and ResourcePool::ExtensionGuide 

  *  Changed Net::LDAP factory to accept the bind parameters in the 
     same way as Net::LDAP expects them. 

  *  Added tests 

V0.9907 - 2002-08-30

  *  Fixed a major bug in Net::LDAP bindings which was introduced in 
     V0.9906. Every LDAP connection was believed to be broken when 
     handed back via free(). 

V0.9906 - 2002-07-10

  *  wrapped DBI calls into eval{} blocks to make RaiseError aware. 

  *  It's possible to pass the options for ResourcePool and 
     LoadBalancer using a Hash reference. 

  *  added tests 

V0.9905 - 2002-07-05

  *  A rollback() is performed on DBI sessions with AutoCommit 
     disabled when they are free()ed. 

  *  free() and fail() for ResourcePool and LoadBalancer return 
     meaningful values. 

  *  added tests 

V0.9904 - 2002-06-04

  *  Implemented the SleepOnFail option for ResourcePool and 
     LoadBalancer. 

  *  Fixed typo in Net::LDAP bindings which caused incomplete error 
     reporting (thx to Sean Welsh) 

  *  Fixed many typo's in the documentation (thx to Florian 
     Helmberger) 

  *  Greatly improved and extended the documentation. 

V0.9903 - 2001-10-15

  *  first public release 



Makefile.PL
#*********************************************************************
#*** Makefile.PL
#*** Copyright (c) 2002,2009 by Markus Winand <mws@fatalmind.com>
#*** $Id: Makefile.PL,v 1.21 2009-11-25 14:40:22 mws Exp $
#*********************************************************************

use ExtUtils::MakeMaker qw(prompt WriteMakefile);
use Data::Dumper;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my %opts;

$opts{NAME}	= 'ResourcePool',
$opts{VERSION} = '1.0106';

if ($ExtUtils::MakeMaker::VERSION >= 5.43) {
    $opts{ABSTRACT} = 'A connection caching and pooling class.';
    $opts{AUTHOR} = 'Markus Winand <mws@fatalmind.com>';
}

if (0) { # interactive testing not finished yet
eval("use Net::LDAP;");
if (exists $INC{"Net/LDAP.pm"}) {
	printf("Great, you have Net::LDAP installed....\n\n");
	eval("use Net::LDAP::Constant qw(:all)");
	printf("If you have a ldap server which can be used for testing,\n");
	printf("please enter the data below, if not just press enter\n");
	my ($ldaph, $newparam);
	do {
		print("Hostname of your ldap server " .
			"(use 'host:port' to use non std. port)\n");
		$ldap_host = prompt("LDAP host:", "");
		if ($ldap_host ne "") {
			printf("Trying to connect via Net::LDAP....");
			my ($host, $port) = split(':', $ldap_host);
			if (defined $port) {
				$newparam = [$host, port => $port];
				print Dumper($newparam);
			} else {
				$newparam = [$host];
			}
			$ldaph = Net::LDAP->new(@{$newparam});
			if (defined $ldaph) {
				printf("looks good\n");
			} else {
				printf("failed\n\nYou can change the hostname or just ".
					"press enter to skipp testing\n");
			}
		}
	} while ($ldap_host ne "" && ! defined $ldaph);	
	if (defined $ldap_host && $ldap_host eq "") {
		undef $ldap_host;
	}
	printf("\n\n");
	if (defined $ldap_host) {
		printf("tying to anonymously bind against %s...", $ldap_host);
		my $rc = $ldaph->bind();
		if ($rc->code == LDAP_SUCCESS) {
			printf("looks good\n");
		} else {
			printf("failed\n");
		}
	}
	
	if (defined $ldap_host) {
		$Data::Dumper::Indent = 0;
		$Data::Dumper::Terse = 1;
		printf("Test configuration\n");	
		printf("==================\n");	
		printf("new() parameters: %s\n", Dumper($newparam));
	}
}

} # disabled
WriteMakefile(%opts);

eval("use ResourcePool;");
eval("use DBI;");
eval("use Net::LDAP;");

if (exists $INC{"ResourcePool.pm"}) {
	my @suggestion;
	if ($ResourcePool::VERSION < 1.0000) {
		print("\n=================================================================================\n");
		printf("\n\nYou are upgrading from ResourcePool %s to %s!\n\n", 
			$ResourcePool::VERSION, $opts{VERSION});
		print( <<PLAIN_TEXT );
The bindings for DBI and Net::LDAP which were previously included
in the ResourcePool core package have been removed from this distribution.

Anyway, they are available as separate distributions on CPAN.

PLAIN_TEXT

		if (exists $INC{"DBI.pm"}) {
			push @suggestion, "ResourcePool::Resource::DBI";
		}
		if (exists $INC{"Net/LDAP.pm"}) {
			push @suggestion, "ResourcePool::Resource::Net::LDAP";
		}
		if (scalar(@suggestion) > 0) {
			print("For this system it looks like you need to install those packages:\n");
			foreach $pack (@suggestion) {
				printf("   -> %s\n", $pack);
			}
		}
		print("\n=================================================================================\n");
	}
}
